### PR TITLE
New events and events corrections

### PR DESCRIPTION
**Pull request by vcoindet**
**Adding the following events :**
- animationstarted
- animationstopped
- zoomchanged

**Improvements on the following events :**
- centerchanged (called through user interaction)
- orienationchanged (called through user interaction, access to oldOrientation)
- layerchanged (access to old and state of the layer)
- layerchanged:visible (access to oldvisibility)
- layerchanged:opacity (access to oldopacity)
- layerchanged:index (access to oldindex)
- layeradded (access to the id of the layer added)

### CODE REVIEW

**Comment by vcoindet** [comment]
Rebased and Ready

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 1]
```diff
@@ -18,17 +18,101 @@ import { STRATEGY_MIN_NETWORK_TRAFFIC } from '../../../../Scene/LayerUpdateStrat
...
+var eventZoom = new CustomEvent('zoomchanged', {
+    detail: {
+        oldZoom: undefined,
+        newZoom: undefined,
+    },
+});
+var _handlerZoom;
+
+/* eslint-disable no-console */
+const logStep = function logStep(step) {
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: remove

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 2]
```diff
@@ -18,17 +18,101 @@ import { STRATEGY_MIN_NETWORK_TRAFFIC } from '../../../../Scene/LayerUpdateStrat
...
-var eventRange = new CustomEvent('rangeChanged');
-var eventCenter = new CustomEvent('centerchanged');
-var eventOrientation = new CustomEvent('orientationchanged');
+var eventRange = new CustomEvent('rangeChanged', {
+    detail: {
+        oldRange: undefined,
+    },
+});
+var eventCenter = new CustomEvent('centerchanged', {
+    detail: {
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: replace by `detail: {old: {},new: {}},`

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 3]
```diff
@@ -156,11 +241,20 @@ ApiGlobe.prototype.moveLayerDown = function moveLayerDown(layerId) {
  * @param      {number}  newIndex   The new index
  */
 ApiGlobe.prototype.moveLayerToIndex = function moveLayerToIndex(layerId, newIndex) {
+    eventLayerChanged.detail.visibility.oldVisibility = this.scene.getMap().layersConfiguration.isColorLayerVisible(layerId);
+    eventLayerChanged.detail.visibility.newVisibility = this.scene.getMap().layersConfiguration.isColorLayerVisible(layerId);
+    eventLayerChanged.detail.opacity.oldOpacity = this.scene.getMap().layersConfiguration.getColorLayerOpacity(layerId);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: factorize

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 4]
```diff
@@ -374,10 +469,18 @@ ApiGlobe.prototype.setRealisticLightingOn = function setRealisticLightingOn(valu
  */
 
 ApiGlobe.prototype.setLayerVisibility = function setLayerVisibility(id, visible) {
+    eventLayerChanged.detail.index.oldIndex = this.scene.getMap().layersConfiguration.layersState[id].sequence;
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: factorize

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 5]
```diff
@@ -389,10 +492,18 @@ ApiGlobe.prototype.setLayerVisibility = function setLayerVisibility(id, visible)
  */
 
 ApiGlobe.prototype.setLayerOpacity = function setLayerOpacity(id, opacity) {
+    eventLayerChanged.detail.index.oldIndex = this.scene.getMap().layersConfiguration.layersState[id].sequence;
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: factorize

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 6]
```diff
@@ -614,14 +735,17 @@ ApiGlobe.prototype.setSceneLoaded = function setSceneLoaded() {
  */
 ApiGlobe.prototype.setCenter = function setCenter(coordinates, isAnimated) {
     isAnimated = isAnimated || this.isAnimationEnabled();
-    eventCenter.oldCenter = this.getCenter();
+    eventCenter.detail.oldCenter.longitude = this.getCenter()._values[0] * 180 / Math.Pi;
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Math.Pi???

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 7]
```diff
@@ -614,14 +735,17 @@ ApiGlobe.prototype.setSceneLoaded = function setSceneLoaded() {
...
+    eventCenter.detail.oldCenter.latitude = this.getCenter()._values[1] * 180 / Math.Pi;
     const position3D = new C.EPSG_4326(coordinates.longitude, coordinates.latitude, 0)
         .as('EPSG:4978').xyz();
     position3D.range = coordinates.range;
     return this.scene.currentControls().setCenter(position3D, isAnimated).then(() => {
         this.scene.notifyChange(1);
         return this.setSceneLoaded().then(() => {
             this.scene.currentControls().updateCameraTransformation();
+            eventCenter.detail.newCenter.longitude = this.getCenter()._values[0] * 180 / Math.Pi;
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: use Coordinates functions

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 8]
```diff
@@ -817,6 +944,109 @@ ApiGlobe.prototype.setZoomScale = function setZoomScale(zoomScale, pitch, isAnim
     return this.setRange(range, isAnimated);
 };
 
+ApiGlobe.prototype.testEvents = function testEvents(eventname, param) {
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: remove and keep in you side

**Review by gchoqueux — CHANGES_REQUESTED** [review]

**Comment by vcoindet** [comment]
rebased and ready @gchoqueux

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 9]
```diff
@@ -11,24 +11,53 @@ import WMTS_Provider from '../../Providers/WMTS_Provider';
...
 import Fetcher from '../../Providers/Fetcher';
 import { STRATEGY_MIN_NETWORK_TRAFFIC } from '../../../../Scene/LayerUpdateStrategy';
 
 var sceneIsLoaded = false;
 var eventLoaded = new CustomEvent('globe-loaded');
-var eventRange = new CustomEvent('rangeChanged');
-var eventCenter = new CustomEvent('centerchanged');
-var eventOrientation = new CustomEvent('orientationchanged');
+var eventRange = new CustomEvent('rangeChanged', {
+    detail: { new: {}, old: {} },
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: create object `statut`
 
detail : new statut(),

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 10]
```diff
@@ -366,6 +403,32 @@ ApiGlobe.prototype.setRealisticLightingOn = function setRealisticLightingOn(valu
     this.scene.renderScene3D();
 };
 
+ApiGlobe.prototype.getOldStatusLayer = function getOldStatusLayer(layerId) {
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Only one function is required `getStatusLayer`, no?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 11]
```diff
@@ -366,6 +403,32 @@ ApiGlobe.prototype.setRealisticLightingOn = function setRealisticLightingOn(valu
...
+    eventLayerChanged.detail.index.old = this.scene.getMap().layersConfiguration.layersState[layerId].sequence;
+    eventLayerChanged.detail.opacity.old = this.scene.getMap().layersConfiguration.getColorLayerOpacity(layerId);
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: `eventLayerChanged.detail.opacity.old` or `eventLayerChangedOpacity.detail.old`
need homogenized details

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 12]
```diff
@@ -843,6 +967,41 @@ ApiGlobe.prototype.callbackLayerChanged = function callbackLayerChanged() {
     this.dispatchEvent(eventLayerChanged);
 };
 
+ApiGlobe.prototype.addEventListenerCenterChanged = function addEventListenerCenterChanged() {
+    this.getCenterStatus('new');
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: and the old statut?

File: src/Renderer/ThreeExtended/GlobeControls.js [hunk 13]
```diff
@@ -651,6 +656,15 @@ function GlobeControls(camera, domElement, engine) {
             return;
         }
 
+        var targt = {};
+        targt.x = globeTarget.position.x;
+        targt.y = globeTarget.position.y;
+
+        var rot = {};
+        rot.x = globeTarget.rotation._x;
```
Comment (gchoqueux) [inline src/Renderer/ThreeExtended/GlobeControls.js]: not correct, i think is more `spherical` member

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 14]
```diff
@@ -865,6 +1033,22 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+    this.viewerDiv.removeEventListener('centercontrolchanged', _handlerCenter, false);
+};
+ApiGlobe.prototype.removeEventListenerZoomChanged = function removeEventListenerZoomChanged() {
+    this.viewerDiv.removeEventListener('zoomcontrolchanged', _handlerZoom, false);
+};
+ApiGlobe.prototype.removeEventListenerOrientationChanged = function removeEventListenerOrientationChanged() {
+    this.viewerDiv.removeEventListener('orientationcontrolchanged', _handlerOrientation, false);
+};
+
+ApiGlobe.prototype.selectNodeById = function selectNodeById(id) {
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Why you add this?

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 12 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: No need here, it's just a initialization so the old status will be good in the function `callbackCenterChanged`

It's the same for the events `zoomchanged` and `orientationchanged`

**Comment by autra** [comment]
2 important points:

- A lot of boilerplate would have been spared if you didn't try to create events as constants... Also doing so is dangerous: sharing them as it is done here (and in itowns currently) is a risk of data leakage and of collision. Just create them when you fire them. Events are not made to be shareable like that. 

It would be simpler and safer and would avoid having stuff like the `status` function, that really does nothing.

*  even if there might be exceptions. ApiGlobe should *not* fire any events : it's proxying stuff to the underlying services (Scene, GlobeControls) etc and *they* should fire events. The point of events is to notify client of changes that itowns is doing without the client knowing *or* to notify them of any asynchronous action that have been completed.

For example: `moveLayerUp` is not asynchrone AND is called *by* the client. The client code (geoportail, or anything else) already *knows* about the change. It simply doesn't make sense to fire an event there. Other examples of useless events : all the setLayer* methods,

**Comment by autra** [comment]
> it's proxying stuff to the underlying services (Scene, GlobeControls) etc and they should fire events

*if* there is an asynchrone process of course.

**Comment by gchoqueux** [comment]
moved in #306 and #298

**Pull request closed without merging**